vcd/config.go

```diff
@@ -90,6 +91,79 @@ func debugPrintf(format string, args ...interface{}) {
 	}
 }
 
+// This is a global MutexKV for all resources
+var vcdMutexKV = mutexkv.NewMutexKV()
+
+func (cli *VCDClient) lockVapp(d *schema.ResourceData) {
+	vappName := d.Get("name").(string)
+	if vappName == "" {
+		panic("vapp name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", cli.getOrgName(d), cli.getVdcName(d), vappName)
+	vcdMutexKV.Lock(key)
+}
+
+func (cli *VCDClient) unLockVapp(d *schema.ResourceData) {
+	vappName := d.Get("name").(string)
+	if vappName == "" {
+		panic("vapp name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", cli.getOrgName(d), cli.getVdcName(d), vappName)
+	vcdMutexKV.Unlock(key)
+}
+
+func (cli *VCDClient) lockParentVapp(d *schema.ResourceData) {
+	vappName := d.Get("vapp_name").(string)
+	if vappName == "" {
+		panic("vapp name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", cli.getOrgName(d), cli.getVdcName(d), vappName)
+	vcdMutexKV.Lock(key)
+}
+
+func (cli *VCDClient) unLockParentVapp(d *schema.ResourceData) {
+	vappName := d.Get("vapp_name").(string)
+	if vappName == "" {
+		panic("vapp name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|vapp:%s", cli.getOrgName(d), cli.getVdcName(d), vappName)
+	vcdMutexKV.Unlock(key)
+}
+
+func (cli *VCDClient) lockParentEdgeGtw(d *schema.ResourceData) {
+	edgeGtwName := d.Get("edge_gateway").(string)
+	if edgeGtwName == "" {
+		panic("edge gtw name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|edge:%s", cli.getOrgName(d), cli.getVdcName(d), edgeGtwName)
+	vcdMutexKV.Lock(key)
+}
+
+func (cli *VCDClient) unLockParentEdgeGtw(d *schema.ResourceData) {
+	edgeGtwName := d.Get("edge_gateway").(string)
+	if edgeGtwName == "" {
+		panic("edge gtw name isn't found")
+	}
+	key := fmt.Sprintf("org:%s|vdc:%s|edge:%s", cli.getOrgName(d), cli.getVdcName(d), edgeGtwName)
+	vcdMutexKV.Unlock(key)
+}
+
+func (cli *VCDClient) getOrgName(d *schema.ResourceData) string {
```

Comment:
What's the purpose of this function?
If the resource doesn't have the Org, we get it from the client? In which cases?
What if the client doesn't have it either?

Reply:
This is the same we have everywhere. Or provider has to have org or resource explicitly state it.
"Inheritance" is that provider org is added to client and that's how we access when resource don't have it. If provider and resource won't define it, then we have errors everywhere. Need to test which part will fail first. It would depend on provider wrong configuration